### PR TITLE
PHPStan: generics on Transatable/Translation

### PR DESCRIPTION
**Pull request by Kocal**
This PR is a proposal for #516. 

The following code:
Capture d’écran de 2020-02-04 07-09-01

Will now be analysed correctly by PHPStan:
Capture d’écran de 2020-02-04 07-09-19

### CODE REVIEW

File: src/Model/Translatable/TranslatableMethodsTrait.php [hunk 1]
```diff
@@ -130,7 +131,7 @@ public static function getTranslationEntityClass(): string
      *
      * @param string $locale The locale (en, ru, fr) | null If null, will try with current locale
      */
-    protected function doTranslate(?string $locale = null, bool $fallbackToDefault = true): TranslationInterface
+    protected function doTranslate(?string $locale = null, bool $fallbackToDefault = true)
```
Comment (Kocal) [inline src/Model/Translatable/TranslatableMethodsTrait.php]: Not a big fan, but it prevents the following issue:

Capture d’écran de 2020-02-04 07-18-40

Also this is a protected method which is only consumed by `translate()` (which still have `TranslationInterface` php return type)

File: src/Model/Translatable/TranslatableMethodsTrait.php [hunk 2]
```diff
@@ -163,7 +164,6 @@ protected function doTranslate(?string $locale = null, bool $fallbackToDefault =
 
         $class = static::getTranslationEntityClass();
 
-        /** @var TranslationInterface $translation */
```
Comment (Kocal) [inline src/Model/Translatable/TranslatableMethodsTrait.php]: Prevent this issue:
Capture d’écran de 2020-02-04 07-20-24


Not a big deal, this modification is inside `doTranslate()` method which is protected

File: src/Model/Translatable/TranslationPropertiesTrait.php [hunk 3]
```diff
@@ -15,7 +15,7 @@ trait TranslationPropertiesTrait
 
     /**
      * Will be mapped to translatable entity by TranslatableSubscriber
-     * @var TranslatableInterface
+     * @var mixed
```
Comment (Kocal) [inline src/Model/Translatable/TranslationPropertiesTrait.php]: Prevent this issue:

Capture d’écran de 2020-02-04 07-21-33

Not a big fan either, but this property is only used by `getTranslatable()` which has:
- php return type to `TranslatableInterface`
- phpstan return type to `TranslatableInterface<Translatable>`

File: phpstan.neon [hunk 4]
```diff
@@ -12,7 +12,19 @@ parameters:
...
+        - '#Call to an undefined method Knp\\DoctrineBehaviors\\Contract\\Provider\\UserProviderInterface::changeUser\(\)#'
+        - '#Call to an undefined method Knp\\DoctrineBehaviors\\Contract\\Entity\\SluggableInterface::getId\(\)#'
+        - '#Call to an undefined method Knp\\DoctrineBehaviors\\Contract\\Entity\\TreeNodeInterface::getId\(\)#'
+        - '#Call to an undefined method Knp\\DoctrineBehaviors\\Contract\\Entity\\TreeNodeInterface::to(Flat)?Array\(\)#'
+
+        # translation proxy
+        - message: '#Call to an undefined method Knp\\DoctrineBehaviors\\Tests\\Fixtures\\Entity\\TranslatableEntity::(get|set)Title\(\)#'
+          path: tests/ORM/TranslatableTest.php
+
+        # translation inheritance, I don't wanted to change (Abstract|Extended)TranslationEntity(Translation)? classes
```
Comment (Kocal) [inline phpstan.neon]: Prevent this issue: 
Capture d’écran de 2020-02-04 07-25-14

Using `@implements` on abstract classes/children classes does not work very well. To make it work, some modifications are necessary but they are not pertinent and wanted I guess:
- `AbstractTranslatableEntity` should not implement `TranslatableInterface` and use trait `TranslatableTrait`
- `AbstractTranslatableEntityTranslation` should not implement `TranslationInterface` and use trait `TranslationTrait`
- `ExtendedTranslatableEntity` should implement `TranslatableInterface`, use trait `TranslatableTrait` and add `@implements` tag
- `ExtendedTranslatableEntityTranslation` should implement `TranslationInterface`, use trait `TranslationTrait` and add `@implements` tag

And so... inheritance now become useless :sweat_smile:

File: src/Model/Translatable/TranslatableMethodsTrait.php [hunk 5]
```diff
@@ -11,9 +11,6 @@
 
 trait TranslatableMethodsTrait
 {
-    /**
-     * @return Collection|TranslationInterface[]
-     */
     public function getTranslations()
```
Comment (TomasVotruba) [inline src/Model/Translatable/TranslatableMethodsTrait.php]: This should be returned, as it weakens PHPStorm autocomplete abilities

File: src/Model/Translatable/TranslatableMethodsTrait.php (hunk 1 above)
Comment (TomasVotruba) [reply src/Model/Translatable/TranslatableMethodsTrait.php]: Don't let tools force you to write bad code just to make them happy.
This destroys contract completely, must be reverted.

File: src/Model/Translatable/TranslatableMethodsTrait.php (hunk 2 above)
Comment (TomasVotruba) [reply src/Model/Translatable/TranslatableMethodsTrait.php]: Must be reverted

File: src/Model/Translatable/TranslationPropertiesTrait.php (hunk 3 above)
Comment (TomasVotruba) [reply src/Model/Translatable/TranslationPropertiesTrait.php]: Same thing, must be reverted to keep contract

File: src/Model/Translatable/TranslatableMethodsTrait.php (hunk 5 above)
Comment (Kocal) [reply src/Model/Translatable/TranslatableMethodsTrait.php]: We already have the PHPDoc on the [TranslatableInterface](https://github.com/Kocal/DoctrineBehaviors/blob/generics/translatable-translation/src/Contract/Entity/TranslatableInterface.php#L14-L18), it does not break PHPStorm autocomplete:
Capture d’écran de 2020-02-11 21-34-12

File: src/Model/Translatable/TranslatableMethodsTrait.php (hunk 1 above)
Comment (Kocal) [reply src/Model/Translatable/TranslatableMethodsTrait.php]: `doTranslate()` is a protected method which is not part of `TranslatableInterface` interface, so the contract is not broken. 
The method is only used by public method `translate()` which still have `TranslatableInterface` return type, so I think it's okay?

File: src/Model/Translatable/TranslatableMethodsTrait.php (hunk 5 above)
Comment (TomasVotruba) [reply src/Model/Translatable/TranslatableMethodsTrait.php]: How about human readability?

**Comment by Kocal** [comment]
Hum, I'm really not proud of this PR. At first I _just_ wanted to add generics support with PHPStan but I had to modifiy too much PHP code...

I will close this PR and open another PR for a PHPStan extension. The final user will be able to use it like this:
```neon
includes:
    - vendor/knplabs/doctrine-behaviors/extension.neon
```

**Pull request closed without merging**